File: vito/core/src/main/java/com/facebook/fresco/vito/core/FrescoController2Impl.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core;

import android.graphics.drawable.Drawable;
import com.facebook.common.references.CloseableReference;
import com.facebook.datasource.DataSource;
import com.facebook.drawee.backends.pipeline.info.ImageOrigin;
import com.facebook.fresco.vito.listener.ImageListener;
import com.facebook.imagepipeline.image.CloseableImage;
import java.util.concurrent.Executor;
import javax.annotation.Nullable;

public class FrescoController2Impl implements DrawableDataSubscriber, FrescoController2 {

  private final Hierarcher mHierarcher;
  private final Executor mLightweightBackgroundThreadExecutor;
  private final Executor mUiThreadExecutor;
  private final VitoImagePipeline mImagePipeline;
  private final @Nullable VitoImageRequestListener mGlobalImageListener;

  public FrescoController2Impl(
      Hierarcher hierarcher,
      Executor lightweightBackgroundThreadExecutor,
      Executor uiThreadExecutor,
      VitoImagePipeline imagePipeline,
      @Nullable VitoImageRequestListener globalImageListener) {
    mHierarcher = hierarcher;
    mLightweightBackgroundThreadExecutor = lightweightBackgroundThreadExecutor;
    mUiThreadExecutor = uiThreadExecutor;
    mImagePipeline = imagePipeline;
    mGlobalImageListener = globalImageListener;
  }

  @Override
  public boolean fetch(
      final FrescoDrawable2 frescoDrawable,
      final VitoImageRequest imageRequest,
      final @Nullable Object callerContext,
      final @Nullable ImageListener listener) {
    // Check if we already fetched the image
    if (frescoDrawable.getDrawableDataSubscriber() == this
        && frescoDrawable.isFetchSubmitted()
        && imageRequest.equals(frescoDrawable.getImageRequest())) {
      frescoDrawable.cancelReleaseNextFrame();
      frescoDrawable.cancelReleaseDelayed();
      return true; // already set
    }
    // We didn't -> Reset everything
    frescoDrawable.close();
    // Basic setup
    frescoDrawable.setDrawableDataSubscriber(this);
    frescoDrawable.setImageRequest(imageRequest);
    frescoDrawable.setCallerContext(callerContext);
    frescoDrawable.setImageListener(listener);
    frescoDrawable.setVitoImageRequestListener(mGlobalImageListener);

    // Set layers that are always visible
    frescoDrawable.setOverlayDrawable(
        mHierarcher.buildOverlayDrawable(imageRequest.resources, imageRequest.imageOptions));

    // We're fetching a new image, so we're updating the ID
    final long imageId = VitoUtils.generateIdentifier();
    frescoDrawable.setImageId(imageId);

    // Notify listeners that we're about to fetch an image
    frescoDrawable.getImageListener().onSubmit(imageId, imageRequest, callerContext);

    // Check if the image is in cache
    CloseableReference<CloseableImage> cachedImage = mImagePipeline.getCachedImage(imageRequest);
    try {
      if (CloseableReference.isValid(cachedImage)) {
        frescoDrawable.setImageOrigin(ImageOrigin.MEMORY_BITMAP);
        // Immediately display the actual image.
        setActualImage(frescoDrawable, imageRequest, cachedImage, true);
        return true;
      }
    } finally {
      CloseableReference.closeSafely(cachedImage);
    }

    // The image is not in cache -> Set up layers visible until the image is available
    frescoDrawable.setProgressDrawable(
        mHierarcher.buildProgressDrawable(imageRequest.resources, imageRequest.imageOptions));
    Drawable placeholder =
        mHierarcher.buildPlaceholderDrawable(imageRequest.resources, imageRequest.imageOptions);
    frescoDrawable.setPlaceholderDrawable(placeholder);
    frescoDrawable.setImageDrawable(null);

    frescoDrawable.getImageListener().onPlaceholderSet(imageId, imageRequest, placeholder);

    // Fetch the image
    mLightweightBackgroundThreadExecutor.execute(
        new Runnable() {
          @Override
          public void run() {
            if (imageId != frescoDrawable.getImageId()) {
              return; // We're trying to load a different image -> ignore
            }
            DataSource<CloseableReference<CloseableImage>> dataSource =
                mImagePipeline.fetchDecodedImage(
                    imageRequest, callerContext, frescoDrawable.getImageOriginListener(), imageId);
            frescoDrawable.setDataSource(dataSource);
            dataSource.subscribe(frescoDrawable, mUiThreadExecutor);
          }
        });
    frescoDrawable.setFetchSubmitted(true);
    return false;
  }

  @Override
  public void releaseDelayed(final FrescoDrawable2 drawable) {
    drawable.scheduleReleaseDelayed();
  }

  @Override
  public void release(final FrescoDrawable2 drawable) {
    drawable.scheduleReleaseNextFrame();
    drawable.getImageListener().onRelease(drawable.getImageId(), drawable.getImageRequest());
  }

  private void setActualImage(
      FrescoDrawable2 drawable,
      VitoImageRequest imageRequest,
      CloseableReference<CloseableImage> image,
      boolean isImmediate) {
    mHierarcher.setupActualImageWrapper(
        drawable.getActualImageWrapper(), imageRequest.imageOptions);
    Drawable actualDrawable =
        mHierarcher.setupActualImageDrawable(
            drawable,
            imageRequest.resources,
            imageRequest.imageOptions,
            image,
            drawable.getActualImageWrapper(),
            isImmediate,
            null);
    drawable
        .getImageListener()
        .onFinalImageSet(
            drawable.getImageId(),
            drawable.getImageRequest(),
            drawable.getImageOrigin(),
            image.get(),
            actualDrawable);
    drawable.setProgressDrawable(null);
  }

  @Override
  public void onNewResult(
      FrescoDrawable2 drawable,
      VitoImageRequest imageRequest,
      DataSource<CloseableReference<CloseableImage>> dataSource) {
    if (dataSource == null || !dataSource.hasResult()) {
      return;
    }

    CloseableReference<CloseableImage> image = dataSource.getResult();
    try {
      if (!CloseableReference.isValid(image)) {
        onFailure(drawable, imageRequest, dataSource);
      } else {
        setActualImage(drawable, imageRequest, image, false);
      }
    } finally {
      CloseableReference.closeSafely(image);
    }
  }

  @Override
  public void onFailure(
      FrescoDrawable2 drawable,
      VitoImageRequest imageRequest,
      DataSource<CloseableReference<CloseableImage>> dataSource) {
    Drawable errorDrawable =
        mHierarcher.buildErrorDrawable(imageRequest.resources, imageRequest.imageOptions);
    drawable.setProgressDrawable(null);
    drawable.setImageDrawable(errorDrawable);
    drawable
        .getImageListener()
        .onFailure(
            drawable.getImageId(),
            drawable.getImageRequest(),
            errorDrawable,
            dataSource.getFailureCause());
  }

  @Override
  public void onProgressUpdate(
      FrescoDrawable2 drawable,
      VitoImageRequest imageRequest,
      DataSource<CloseableReference<CloseableImage>> dataSource) {
    // TODO: implement
  }
}

File: vito/core-java-impl/src/main/java/com/facebook/fresco/vito/core/impl/FrescoController2Impl.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core.impl

import android.graphics.Bitmap
import android.graphics.Rect
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import android.net.Uri
import com.facebook.common.callercontext.ContextChain
import com.facebook.common.internal.ImmutableMap
import com.facebook.common.internal.Supplier
import com.facebook.common.logging.FLog
import com.facebook.common.references.CloseableReference
import com.facebook.datasource.DataSource
import com.facebook.drawee.backends.pipeline.info.ImageOrigin
import com.facebook.fresco.middleware.HasExtraData
import com.facebook.fresco.middleware.MiddlewareUtils.obtainExtras
import com.facebook.fresco.ui.common.ControllerListener2.Extras
import com.facebook.fresco.ui.common.ImagePerfDataListener
import com.facebook.fresco.ui.common.ImagePerfDataNotifier
import com.facebook.fresco.ui.common.ImagePerfNotifier
import com.facebook.fresco.ui.common.OnFadeListener
import com.facebook.fresco.ui.common.VitoUtils.generateIdentifier
import com.facebook.fresco.vito.core.FrescoController2
import com.facebook.fresco.vito.core.FrescoDrawableInterface
import com.facebook.fresco.vito.core.FrescoVitoConfig
import com.facebook.fresco.vito.core.ImagePerfLoggingListener
import com.facebook.fresco.vito.core.NopDrawable
import com.facebook.fresco.vito.core.VitoImagePerfListener
import com.facebook.fresco.vito.core.VitoImagePipeline
import com.facebook.fresco.vito.core.VitoImageRequest
import com.facebook.fresco.vito.core.VitoImageRequestListener
import com.facebook.fresco.vito.core.impl.debug.DebugOverlayFactory2
import com.facebook.fresco.vito.listener.ImageListener
import com.facebook.fresco.vito.source.BitmapImageSource
import com.facebook.fresco.vito.source.DrawableImageSource
import com.facebook.fresco.vito.source.EmptyImageSource
import com.facebook.imagepipeline.image.CloseableBitmap
import com.facebook.imagepipeline.image.CloseableImage
import com.facebook.imagepipeline.image.CloseableStaticBitmap
import com.facebook.imagepipeline.image.ImageInfoImpl
import com.facebook.imagepipeline.image.ImmutableQualityInfo
import java.util.concurrent.Executor

open class FrescoController2Impl(
    private val config: FrescoVitoConfig,
    private val hierarcher: Hierarcher,
    private val lightweightBackgroundThreadExecutor: Executor,
    private val uiThreadExecutor: Executor,
    private val imagePipeline: VitoImagePipeline,
    private val globalImageListener: VitoImageRequestListener?,
    private val debugOverlayFactory: DebugOverlayFactory2,
    private val imagePerfListenerSupplier: Supplier<ImagePerfLoggingListener>?,
    private val vitoImagePerfListener: VitoImagePerfListener
) : DrawableDataSubscriber, FrescoController2 {

  @Suppress("UNCHECKED_CAST")
  override fun <T> createDrawable(uiFramework: String?): T where
  T : Drawable,
  T : FrescoDrawableInterface =
      FrescoDrawable2Impl(
          config.useNewReleaseCallback(),
          imagePerfListenerSupplier?.get(),
          vitoImagePerfListener,
          uiFramework,
          config.experimentalResetVitoImageRequestListener(),
          config.experimentalResetLocalVitoImageRequestListener(),
          config.experimentalResetLocalImagePerfStateListener(),
          config.experimentalResetControllerListener2())
          as T

  override fun fetch(
      drawable: FrescoDrawableInterface,
      imageRequest: VitoImageRequest,
      callerContext: Any?,
      contextChain: ContextChain?,
      listener: ImageListener?,
      perfDataListener: ImagePerfDataListener?,
      onFadeListener: OnFadeListener?,
      viewportDimensions: Rect?,
      vitoImageRequestListener: VitoImageRequestListener?
  ): Boolean {
    if (drawable !is FrescoDrawable2Impl) {
      FLog.e(TAG, "Drawable not supported $drawable")
      return false
    }
    // Fast path for null-URIs
    if (config.fastPathForEmptyRequests() && imageRequest.imageSource is EmptyImageSource) {
      emptyRequestFastPath(drawable, imageRequest, callerContext)
      return true
    }

    // Save viewport dimension for future use
    drawable.viewportDimensions = viewportDimensions

    // Check if we already fetched the image
    if (isAlreadyLoadingImage(drawable, imageRequest)) {
      drawable.cancelReleaseNextFrame()
      drawable.cancelReleaseDelayed()
      return true // already set
    }
    if (drawable.isFetchSubmitted) {
      drawable.imagePerfListener.onDrawableReconfigured(drawable)
    }
    // We didn't -> Reset everything
    drawable.close()
    // Basic setup
    drawable.drawableDataSubscriber = this
    drawable.imageRequest = imageRequest
    drawable.callerContext = callerContext
    drawable.imageListener = listener
    drawable.setVitoImageRequestListener(globalImageListener)
    drawable.setLocalVitoImageRequestListener(vitoImageRequestListener)

    // Setup local perf data listener
    if (perfDataListener != null) {
      val localPerfStateListener: ImagePerfNotifier = ImagePerfDataNotifier(perfDataListener)
      drawable.internalListener.setLocalImagePerfStateListener(localPerfStateListener)
    } else {
      drawable.internalListener.setLocalImagePerfStateListener(null)
    }
    drawable.setOnFadeListener(onFadeListener)

    // Set layers that are always visible
    drawable.setOverlayDrawable(
        hierarcher.buildOverlayDrawable(imageRequest.resources, imageRequest.imageOptions))

    // We're fetching a new image, so we're updating the ID
    val imageId = generateIdentifier()
    drawable.imageId = imageId
    val extras = obtainExtras(null, null, drawable, imageRequest)

    // Notify listeners that we're about to fetch an image
    drawable.internalListener.onSubmit(imageId, imageRequest, callerContext, extras)
    drawable.imagePerfListener.onImageFetch(drawable)

    // Direct bitmap available
    if (imageRequest.imageSource is BitmapImageSource) {
      val bitmap = (imageRequest.imageSource as BitmapImageSource).bitmap
      val closeableBitmap: CloseableBitmap =
          CloseableStaticBitmap.of(bitmap, { _: Bitmap? -> }, ImmutableQualityInfo.FULL_QUALITY, 0)
      val bitmapRef = CloseableReference.of<CloseableImage>(closeableBitmap)
      return try {
        drawable.imageOrigin = ImageOrigin.MEMORY_BITMAP
        // Immediately display the actual image.
        setActualImage(drawable, imageRequest, bitmapRef, true, null)
        drawable.setFetchSubmitted(true)
        debugOverlayFactory.update(drawable, extras)
        true
      } finally {
        CloseableReference.closeSafely(bitmapRef)
      }
    } else if (imageRequest.imageSource is DrawableImageSource) {
      val actualImageDrawable = (imageRequest.imageSource as DrawableImageSource).drawable
      val actualImageWrapperDrawable = drawable.actualImageWrapper
      hierarcher.setupActualImageWrapper(
          actualImageWrapperDrawable, imageRequest.imageOptions, drawable.callerContext)
      val actualDrawable =
          hierarcher.applyRoundingOptions(
              imageRequest.resources, actualImageDrawable, imageRequest.imageOptions)
      actualImageWrapperDrawable.setCurrent(actualDrawable)
      drawable.setImage(actualImageWrapperDrawable, null)
      drawable.showImageImmediately()
      drawable.imageOrigin = ImageOrigin.LOCAL
      drawable.setFetchSubmitted(true)
      if (imageRequest.imageOptions.shouldAutoPlay() && actualImageDrawable is Animatable) {
        (actualDrawable as Animatable).start()
      }
      val imageInfoExtras: Map<String, Any> = extras.imageExtras ?: HashMap()
      drawable.internalListener.onFinalImageSet(
          drawable.imageId,
          imageRequest,
          ImageOrigin.LOCAL,
          ImageInfoImpl(
              actualImageDrawable.intrinsicWidth,
              actualImageDrawable.intrinsicHeight,
              0,
              ImmutableQualityInfo.FULL_QUALITY,
              imageInfoExtras),
          extras,
          actualDrawable)
      drawable.imagePerfListener.onImageSuccess(drawable, true)
      debugOverlayFactory.update(drawable, extras)
      return true
    }

    // Check if the image is in cache
    val cachedImage = imagePipeline.getCachedImage(imageRequest)
    var needsPlaceholderDrawable = true
    try {
      if (CloseableReference.isValid(cachedImage)) {
        drawable.imageOrigin = ImageOrigin.MEMORY_BITMAP_SHORTCUT
        val isIntermediateImage =
            config.useIntermediateImagesAsPlaceholder() &&
                cachedImage?.get()?.qualityInfo?.isOfFullQuality != true
        // Immediately display the actual image.
        setActualImage(drawable, imageRequest, cachedImage, true, null, isIntermediateImage)
        drawable.setFetchSubmitted(true)
        debugOverlayFactory.update(drawable, extras)
        if (!isIntermediateImage) {
          return true
        } else {
          // We are displaying an intermediate image and can skip setting up the placeholder
          needsPlaceholderDrawable = false
        }
      }
    } finally {
      CloseableReference.closeSafely(cachedImage)
    }

    if (needsPlaceholderDrawable) {
      // The image is not in cache -> Set up layers visible until the image is available
      drawable.setProgressDrawable(
          hierarcher.buildProgressDrawable(imageRequest.resources, imageRequest.imageOptions))
      // Immediately show the progress image and set progress to 0
      drawable.setProgress(0f)
      drawable.showProgressImmediately()
      setUpPlaceholder(drawable, imageRequest, imageId)
    }

    // Fetch the image
    val fetchRunnable = Runnable {
      if (imageId != drawable.imageId) {
        return@Runnable // We're trying to load a different image -> ignore
      }
      val dataSource =
          imagePipeline.fetchDecodedImage(
              imageRequest, callerContext, drawable.imageOriginListener, imageId)
      drawable.setDataSource(imageId, dataSource)
      dataSource.subscribe(drawable, uiThreadExecutor)
    }
    if (config.submitFetchOnBgThread()) {
      lightweightBackgroundThreadExecutor.execute(fetchRunnable)
    } else {
      fetchRunnable.run()
    }
    drawable.setFetchSubmitted(true)
    debugOverlayFactory.update(drawable, null)
    return false
  }

  private fun emptyRequestFastPath(
      frescoDrawable: FrescoDrawable2Impl,
      imageRequest: VitoImageRequest,
      callerContext: Any?
  ) {
    frescoDrawable.close()
    frescoDrawable.setVitoImageRequestListener(globalImageListener)
    frescoDrawable.internalListener.onEmptyEvent(callerContext)
    frescoDrawable.setOverlayDrawable(
        hierarcher.buildOverlayDrawable(imageRequest.resources, imageRequest.imageOptions))
    setUpPlaceholder(frescoDrawable, imageRequest, EMPTY_IMAGE_ID)
  }

  private fun setUpPlaceholder(
      frescoDrawable: FrescoDrawable2Impl,
      imageRequest: VitoImageRequest,
      imageId: Long
  ) {
    val placeholder =
        hierarcher.buildPlaceholderDrawable(imageRequest.resources, imageRequest.imageOptions)
    frescoDrawable.setPlaceholderDrawable(placeholder)
    frescoDrawable.setImageDrawable(null)
    frescoDrawable.internalListener.onPlaceholderSet(imageId, imageRequest, placeholder)
  }

  override fun releaseDelayed(drawable: FrescoDrawableInterface) {
    if (drawable !is FrescoDrawable2Impl) {
      FLog.e(TAG, "Drawable not supported $drawable")
      return
    }
    drawable.imagePerfListener.onScheduleReleaseDelayed(drawable)
    drawable.scheduleReleaseDelayed()
  }

  override fun release(drawable: FrescoDrawableInterface) {
    if (drawable !is FrescoDrawable2Impl) {
      FLog.e(TAG, "Drawable not supported $drawable")
      return
    }
    drawable.imagePerfListener.onScheduleReleaseNextFrame(drawable)
    drawable.scheduleReleaseNextFrame()
  }

  override fun releaseImmediately(drawable: FrescoDrawableInterface) {
    if (drawable !is FrescoDrawable2Impl) {
      FLog.e(TAG, "Drawable not supported $drawable")
      return
    }
    drawable.imagePerfListener.onReleaseImmediately(drawable)
    drawable.releaseImmediately()
  }

  private fun setActualImage(
      drawable: FrescoDrawable2Impl,
      imageRequest: VitoImageRequest,
      image: CloseableReference<CloseableImage>?,
      displayImmediately: Boolean,
      dataSource: DataSource<CloseableReference<CloseableImage>>?,
      isIntermediateImage: Boolean = false,
  ) {
    val actualImageWrapperDrawable = drawable.actualImageWrapper
    hierarcher.setupActualImageWrapper(
        actualImageWrapperDrawable, imageRequest.imageOptions, drawable.callerContext)
    val actualDrawable =
        image?.let {
          hierarcher.buildActualImageDrawable(imageRequest.resources, imageRequest.imageOptions, it)
        }

    actualImageWrapperDrawable.setCurrent(actualDrawable ?: NopDrawable)
    drawable.setImage(actualImageWrapperDrawable, image)
    if (displayImmediately || imageRequest.imageOptions.fadeDurationMs <= 0) {
      drawable.showImageImmediately()
    } else {
      drawable.fadeInImage(imageRequest.imageOptions.fadeDurationMs)
    }
    if (imageRequest.imageOptions.shouldAutoPlay() && actualDrawable is Animatable) {
      (actualDrawable as Animatable).start()
    }
    val extras = obtainExtras(dataSource, image, drawable, imageRequest)
    val imageInfo = image?.get()?.imageInfo
    if (!isIntermediateImage && notifyFinalResult(dataSource)) {
      drawable.internalListener.onFinalImageSet(
          drawable.imageId, imageRequest, drawable.imageOrigin, imageInfo, extras, actualDrawable)
    } else {
      drawable.internalListener.onIntermediateImageSet(drawable.imageId, imageRequest, imageInfo)
    }
    drawable.imagePerfListener.onImageSuccess(drawable, displayImmediately)
    var progress = 1f
    if (dataSource != null && !dataSource.isFinished) {
      progress = dataSource.progress
    }
    drawable.setProgress(progress)
    debugOverlayFactory.update(drawable, extras)
  }

  override fun onNewResult(
      drawable: FrescoDrawable2Impl,
      imageRequest: VitoImageRequest,
      dataSource: DataSource<CloseableReference<CloseableImage>>
  ) {
    if (!dataSource.hasResult()) {
      return
    }
    val image = dataSource.result
    try {
      if (!CloseableReference.isValid(image)) {
        onFailure(drawable, imageRequest, dataSource)
      } else {
        setActualImage(drawable, imageRequest, image, false, dataSource)
      }
    } finally {
      CloseableReference.closeSafely(image)
    }
  }

  override fun onFailure(
      drawable: FrescoDrawable2Impl,
      imageRequest: VitoImageRequest,
      dataSource: DataSource<CloseableReference<CloseableImage>>
  ) {
    val errorDrawable =
        hierarcher.buildErrorDrawable(imageRequest.resources, imageRequest.imageOptions)
    drawable.setProgress(1f)
    drawable.setImageDrawable(errorDrawable)
    if (!drawable.isDefaultLayerIsOn) {
      if (imageRequest.imageOptions.fadeDurationMs <= 0) {
        drawable.showImageImmediately()
      } else {
        drawable.fadeInImage(imageRequest.imageOptions.fadeDurationMs)
      }
    } else {
      drawable.setPlaceholderDrawable(null)
      drawable.setProgressDrawable(null)
    }
    val extras = obtainExtras(dataSource, dataSource.result, drawable, imageRequest)
    if (notifyFinalResult(dataSource)) {
      drawable.internalListener.onFailure(
          drawable.imageId, imageRequest, errorDrawable, dataSource.failureCause, extras)
    } else {
      drawable.internalListener.onIntermediateImageFailed(
          drawable.imageId, imageRequest, dataSource.failureCause)
    }
    drawable.imagePerfListener.onImageError(drawable)
    debugOverlayFactory.update(drawable, extras)
  }

  override fun onProgressUpdate(
      drawable: FrescoDrawable2Impl,
      imageRequest: VitoImageRequest,
      dataSource: DataSource<CloseableReference<CloseableImage>>
  ) {
    if (!dataSource.isFinished) {
      drawable.setProgress(dataSource.progress)
    }
  }

  override fun onRelease(drawable: FrescoDrawable2Impl) {
    val imageRequest = drawable.imageRequest
    if (imageRequest != null) {
      // Notify listeners
      drawable.internalListener.onRelease(
          drawable.imageId, imageRequest, obtainExtras(null, null, drawable, imageRequest))
    }
    drawable.imagePerfListener.onImageRelease(drawable)
  }

  private fun isAlreadyLoadingImage(
      drawable: FrescoDrawable2Impl,
      imageRequest: VitoImageRequest
  ): Boolean {
    if (drawable.drawableDataSubscriber !== this || !drawable.isFetchSubmitted) {
      return false
    }
    return if (config.useSmartPropertyDiffing()) {
      imageRequest.equalsIfHasImage(drawable.imageRequest, drawable.hasImage())
    } else {
      imageRequest == drawable.imageRequest
    }
  }

  companion object {
    private val COMPONENT_EXTRAS = ImmutableMap.of<String, Any>("component_tag", "vito2")
    private val SHORTCUT_EXTRAS =
        ImmutableMap.of<String, Any>("origin", "memory_bitmap", "origin_sub", "shortcut")
    private const val TAG = "FrescoController2Impl"
    private const val EMPTY_IMAGE_ID = Long.MAX_VALUE

    private fun obtainExtras(
        dataSource: DataSource<CloseableReference<CloseableImage>>?,
        image: CloseableReference<CloseableImage>?,
        drawable: FrescoDrawable2,
        imageRequest: VitoImageRequest
    ): Extras {
      var imageExtras: Map<String, Any>? = null
      if (image != null) {
        imageExtras = image.get().extras
      }
      var sourceUri: Uri? = null
      val vitoImageRequest = drawable.imageRequest
      var imageSourceExtras: Map<String, Any>? = null
      var logWithHighSamplingRate = false
      if (vitoImageRequest != null) {
        logWithHighSamplingRate = vitoImageRequest.logWithHighSamplingRate
        vitoImageRequest.finalImageRequest?.let { finalImageRequest ->
          sourceUri = finalImageRequest.sourceUri
        }
        @Suppress("UNCHECKED_CAST")
        imageSourceExtras =
            vitoImageRequest.extras[HasExtraData.KEY_IMAGE_SOURCE_EXTRAS] as? Map<String, Any>
      }
      val extras =
          obtainExtras(
              COMPONENT_EXTRAS,
              SHORTCUT_EXTRAS,
              dataSource?.extras,
              imageSourceExtras,
              drawable.viewportDimensions,
              drawable.actualImageScaleType.toString(),
              drawable.actualImageFocusPoint,
              imageExtras,
              drawable.callerContext,
              logWithHighSamplingRate,
              sourceUri)
      extras.modifiedUriStatus = imageRequest.extras[HasExtraData.KEY_MODIFIED_URL] as? String
      extras.originalUri = imageRequest.extras[HasExtraData.KEY_ORIGINAL_URL] as? Uri
      extras.uiFramework = drawable.uiFramework
      return extras
    }

    private fun notifyFinalResult(
        dataSource: DataSource<CloseableReference<CloseableImage>>?
    ): Boolean = dataSource == null || dataSource.isFinished || dataSource.hasMultipleResults()
  }
}
